CAS/Chemical mapping management

mcumcu opened the pull request:
NOTE: requires backend update,
see: https://github.com/bitzesty/trade-tariff-backend/pull/328
"Chemical/CAS mapping - admin/chemical/* endpoints"

Add Admin UI to view, edit and add CAS number mappings. Add feature specs to cover new functionality.

Trello: https://trello.com/c/8WjkHQlC/868-admin-ui-to-view-edit-and-add-cas-number-mappings
Asana: https://app.asana.com/0/1199180449660757/1198723049518172/f

Screenshots:

`/chemicals`:
![image](https://user-images.githubusercontent.com/740379/101820514-b5b68100-3af4-11eb-9652-132da1b1c336.png)

`/chemicals/:chemical_id/edit`:
![image](https://user-images.githubusercontent.com/740379/101820591-d41c7c80-3af4-11eb-9c16-044bfc04a3f6.png)

`/chemicals/:chemical_id/map/edit`:
![image](https://user-images.githubusercontent.com/740379/101820680-f31b0e80-3af4-11eb-92a1-9bfd6ccc7a00.png)

`/chemicals/new`:
![image](https://user-images.githubusercontent.com/740379/101820722-0332ee00-3af5-11eb-9d0d-b3f7ff56c645.png)

Pagination added to `/chemicals` in commit bf27fbb:
![image](https://user-images.githubusercontent.com/740379/100287607-f3d37280-2f42-11eb-9a85-c6a8d453ccca.png)

Commit:
✨ CAS/Chemical mapping management

Add Admin UI to view, edit and add CAS number mappings

https://trello.com/c/8WjkHQlC/868-admin-ui-to-view-edit-and-add-cas-number-mappings

Commit:
Skip a feature spec: update chemical mapping

matthewford commented:
@mcumcu for UI changes it would be good to also include a screenshot

alexey-klimuk commented on app/models/chemical.rb:
```diff
@@ -0,0 +1,31 @@
+    end
+    Chemical.find(id)
+  end
+end
```
new line

mcumcu replied:
👍

alexey-klimuk commented on app/views/chemicals/index.html.erb:
```diff
@@ -0,0 +1,27 @@
+      </tr>
+    <% end %>
+  </tbody>
+</table>
```
new line

mcumcu replied:
👍

alexey-klimuk commented on app/views/chemicals/new_map.html.erb:
```diff
@@ -0,0 +1,35 @@
+    <% end %>
+    <%#= link_to "Back", :back, class: 'btn' %>
+  </div>
+<% end %>
```
new line

Commit:
Update specs

Commit:
Adding Chemical create and update features

Commit:
Refactor
- for UI/UX improvements
- rename some parameters

Commit:
Add specs for chemical#create and chemical#update

Commit:
Move chemical map "Delete" button into the "update" form
- modify the controller to detect which button was pressed [update|delete] and act appropriately
- added some newlines
- update UI labels

Commit:
Revise the order of the links in the navbar
- make "CAS" come immediately after "Footnotes"

Commit:
Refactor
- move common code into controller method

alexey-klimuk commented:
@mcumcu is this consistent with the latest api changes on the backend?

alexey-klimuk commented on app/models/chemical_name.rb:
```diff
@@ -0,0 +1,10 @@
+  attributes :id, :name
+
+  belongs_to :chemical
+end
```
newline

alexey-klimuk commented on spec/factories/chemical_factory.rb:
```diff
@@ -0,0 +1,8 @@
+    sequence(:name) { |n| "chemical-#{n}" }
+    goods_nomenclature_map { {} }
+  end
+end
```
newline

alexey-klimuk commented:
@mcumcu could you please separate chemical editing and assigning new goods_nomenclature?
<img width="928" alt="Screenshot 2020-11-24 at 20 48 23" src="https://user-images.githubusercontent.com/1684845/100143993-6e928400-2e96-11eb-986c-8493bc1ef672.png">

The page looks a bit too big for me. @matthewford what do you think? Should we ask for UX help?

matthewford commented:
Not sure how much it will be used to warrant a lot of work on the UI but if
we can make it simpler that would be great, perhaps some sub navigation and
separate pages

On Tue, 24 Nov 2020 at 19:50, Alexey Klimuk <notifications@github.com>
wrote:

> @mcumcu <https://github.com/mcumcu> could you please separate chemical
> editing and assigning new goods_nomenclature?
> [image: Screenshot 2020-11-24 at 20 48 23]
> <https://user-images.githubusercontent.com/1684845/100143993-6e928400-2e96-11eb-986c-8493bc1ef672.png>
>
> The page looks a bit too big for me. @matthewford
> <https://github.com/matthewford> what do you think? Should we ask for UX
> help?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/bitzesty/trade-tariff-admin/pull/117#issuecomment-733196709>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAABVYM7ME72SGWWLE47M3SRQE7VANCNFSM4TJVBRAA>
> .
>
-- 
Sent from my iPhone

-- 
Bit Zesty Ltd, a company incorporated in England & Wales with registered 
number 06883289. Our registered address is 94 White Lion Street, London, N1 
9PF.

Commit:
Add pagination to `/chemicals`

Commit:
Properly report API errors
- when backend API produces errors, make thge Admin UI display a flash message

Commit:
Add newlines

Commit:
Simplify the Chemicals admin UI
- create new routes
- move "chemical editing" and "mapping editing" functions onto separate pages

mcumcu commented:
> Not sure how much it will be used to warrant a lot of work on the UI but if we can make it simpler that would be great, perhaps some sub navigation and separate pages

> @mcumcu could you please separate chemical editing and assigning new goods_nomenclature?

> The page looks a bit too big for me. @matthewford what do you think? Should we ask for UX help?

UI has been simplified, moved GN mapping and chemical editing onto separate pages, see updated screenshots

matthewford commented:
Engine will merge
